[Access] Enable Event streaming on REST API

UlyanaAndrukhiv opened the pull request:
https://github.com/onflow/flow-go/issues/4379

This pull request:

- added a new route type `wsroute` with `SubscribeEvents` and its handler implementation to existing REST router, 
- added new WSHandler to handle websockets requests,
- refactored REST `Handler` by moving common logic to `HttpHandler`, 
- added unit and integration tests

peterargue reviewed (commented):
overall this approach looks good. I added a few comments, but didn't review thoroughly since this is still a work in progress.

durkmurder commented on engine/common/state_stream/subscription_test.go:
```diff
@@ -3,14 +3,15 @@ package state_stream_test
 import (
 	"context"
 	"fmt"
+
```
```suggestion
```

durkmurder commented on engine/common/state_stream/streamer_test.go:
```diff
@@ -3,6 +3,7 @@ package state_stream_test
 import (
 	"context"
 	"fmt"
+
```
```suggestion
```

durkmurder commented on module/state_synchronization/requester/execution_data_requester_test.go:
```diff
@@ -3,6 +3,7 @@ package requester_test
 import (
 	"context"
 	"fmt"
+
```
```suggestion
```

durkmurder commented on cmd/observer/node_builder/observer_builder.go:
```diff
@@ -973,6 +975,9 @@ func (builder *ObserverServiceBuilder) enqueueRPCServer() {
 			restHandler,
 			builder.secureGrpcServer,
 			builder.unsecureGrpcServer,
+			nil,
```
```suggestion
			nil, // state streaming is not supported
```

durkmurder commented on engine/access/rest/middleware/logging.go:
```diff
@@ -40,6 +43,9 @@ type responseWriter struct {
 	statusCode int
 }
 
+// http.Hijacker necessary for upgrading gorilla websocket connection for "subscribe_events" route.
```
```suggestion
// http.Hijacker necessary for using middleware with gorilla websocket connections.
```

durkmurder commented on engine/access/rest/request/event_type.go:
```diff
@@ -0,0 +1,58 @@
+type EventTypes []EventType
+
+func (e *EventTypes) Parse(raw []string) error {
+	if len(raw) > MaxIDsLength {
```
I think we should comply to `EventFilterConfig.MaxEventTypes` or something similar in this. I would say we should mirror as much as possible GRPC logic.

durkmurder replied:
Can we actually rely on `NewEventFilter` to do the validation? It seems it already has checks for event types, contracts and addresses + it has overflow checks.

durkmurder commented on engine/access/rest/routes/accounts_test.go:
```diff
@@ -2,6 +2,7 @@ package routes
 
 import (
 	"fmt"
+
```
```suggestion
```

durkmurder commented on engine/access/rest/request/subscribe_events.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	g.EventTypes = eventTypes.Flow()
+
```
```suggestion
```

durkmurder commented on engine/access/rest/routes/accounts_test.go:
```diff
@@ -44,6 +45,7 @@ func accountURL(t *testing.T, address string, height string) string {
 //	5. Get invalid account.
 func TestAccessGetAccount(t *testing.T) {
 	backend := &mock.API{}
+	stateStreamBackend := &mock_state_stream.API{}
```
```suggestion
	stateStreamBackend := mock_state_stream.NewAPI(t)
```

durkmurder replied:
Since state stream backend can be nil, can we pass nil in this tests?

durkmurder commented on engine/access/rest/routes/blocks_test.go:
```diff
@@ -143,13 +144,14 @@ func prepareTestVectors(t *testing.T,
 
 	for _, tv := range testVectors {
-		responseRec, err := executeRequest(tv.request, backend)
+		responseRec, err := executeRequest(tv.request, backend, stateStreamBackend)
```
Can you just pass nil here? We don't need state streaming for this request

durkmurder commented on engine/access/rest/routes/blocks_test.go:
```diff
@@ -143,13 +144,14 @@ func prepareTestVectors(t *testing.T,
 // TestGetBlocks tests local get blocks by ID and get blocks by heights API
 func TestAccessGetBlocks(t *testing.T) {
 	backend := &mock.API{}
+	stateStreamBackend := &mock_state_stream.API{}
```
```suggestion
	stateStreamBackend := mock_state_stream.NewAPI(t)
```

durkmurder commented on cmd/access/node_builder/access_node_builder.go:
```diff
@@ -616,20 +619,33 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionDataRequester() *FlowAccessN
 				builder.executionDataConfig.InitialBlockHeight,
-				highestAvailableHeight,
+				highestAvailableHeight)
+
```
```suggestion
```

durkmurder commented on engine/access/rest/server.go:
```diff
@@ -9,13 +9,23 @@ import (
+	maxGlobalStreams uint32,
+) (*http.Server, error) {
+	router, err := routes.NewRouter(serverAPI, logger, chain, restCollector, stateStreamApi, eventFilterConfig, maxGlobalStreams)
+
```
```suggestion
```

durkmurder commented on engine/access/rest/routes/handler.go:
```diff
@@ -47,31 +37,26 @@ func NewHandler(
 		linkGenerator:  generator,
-		chain:          chain,
 	}
+	handler.HttpHandler = NewHttpHandler(logger, chain)
```
can you merge initialization of `HttpHandler` in statement above?

peterargue reviewed (commented):
This is coming together nicely! Added a few comments. Mostly focused on the server synchronization since that's the trickiest part.

peterargue commented on integration/tests/access/rest_state_stream_test.go:
```diff
@@ -0,0 +1,242 @@
+		// Wait for events or timeout
+		for {
+			select {
+			case <-timeoutChan:
```
we don't really need a channel
```suggestion
			case <-time.After(10 * time.Second):
```

peterargue commented on integration/tests/access/rest_state_stream_test.go:
```diff
@@ -0,0 +1,242 @@
+			select {
+			case <-timeoutChan:
+				// Handle the timeout and close the client connection
+				client.Close()
```
put this in a defer immediately after creating the client. that way it's easy to see it's done and harder to accidentally miss a case later when adding to this test.

peterargue commented on integration/tests/access/rest_state_stream_test.go:
```diff
@@ -0,0 +1,242 @@
+		}()
+
+		// Wait for events or timeout
+		for {
```
I'm not totally clear on what the expected behavior is here. There seems to be 2 valid cases:
1. Test runs for 10 seconds and then all events from each response received in that time is compared against a different API call made to the same AN
2. The test received at least one event. response and the server shutdown unexpectedly.

I don't think both could be valid results. Let's pick one. Either, check that the server returns at least one response with events then end the test. Or, expect it to run for the full 10 time period.

peterargue commented on engine/access/rest/routes/subscribe_events_test.go:
```diff
@@ -0,0 +1,343 @@
+			startBlockFound := test.startBlockID == flow.ZeroID
+
+			// Helper function to check if a string is present in a slice
+			addExpectedEvent := func(slice []string, item string) bool {
```
would [`slices.Contains`](https://pkg.go.dev/golang.org/x/exp@v0.0.0-20230321023759-10a507213a29/slices#Contains) work?

peterargue reviewed (commented):
Thanks for the revisions! this version is looking pretty good. mostly small comments. I still need to finish reviewing the tests, but wanted to provide some feedback earlier

peterargue reviewed (commented):
Thanks for the revisions! this version is looking pretty good. mostly small comments. I still need to finish reviewing the tests, but wanted to provide some feedback earlier

UlyanaAndrukhiv commented on engine/access/rest/routes/websocket_handler.go:
```diff
@@ -0,0 +1,269 @@
+				wsController.wsErrorHandler(models.NewRestError(http.StatusInternalServerError, "Set the initial write deadline error: ", err))
+				return
+			}
+			if err := wsController.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
```
yes, cause the ping is used to check if the connection is still alive

durkmurder commented on engine/access/rest/routes/test_helpers.go:
```diff
@@ -24,28 +32,112 @@ const (
 )
 
-func executeRequest(req *http.Request, backend access.API) (*httptest.ResponseRecorder, error) {
+type fakeNetConn struct {
```
```suggestion
// fakeNetConn implements a mocked ws connection that can be injected in testing logic.
type fakeNetConn struct {
```

durkmurder commented on engine/access/rest/routes/test_helpers.go:
```diff
@@ -24,28 +32,112 @@ const (
+	closed chan struct{}
+}
+
+// Close closes the fakeNetConn and signals its closure by closing the "closed" channel.
```
```suggestion
var _ net.Conn = (*fakeNetConn)(nil)

// Close closes the fakeNetConn and signals its closure by closing the "closed" channel.
```

durkmurder commented on engine/access/rest/routes/websocket_handler.go:
```diff
@@ -0,0 +1,279 @@
+	for {
+		select {
+		case _, ok := <-wsController.readChannel:
+			if !ok {
```
```suggestion
                        // we use `readChannel` as indicator of client's status, when `readChannel` closes it means that client connection has been terminated and we need to stop this goroutine to avoid memory leak.
			if !ok {
```

durkmurder commented on engine/access/rest/routes/websocket_handler.go:
```diff
@@ -0,0 +1,279 @@
+		api:               api,
+		eventFilterConfig: eventFilterConfig,
+		maxStreams:        int32(maxGlobalStreams),
+		streamCount:       &atomic.Int32{},
```
```suggestion
		streamCount:       atomic.NewInt32(0),
```

we usually use "go.uber.org/atomic" instead of "sync/atomic"

durkmurder commented on engine/access/rest/routes/websocket_handler.go:
```diff
@@ -0,0 +1,279 @@
+	api               state_stream.API
+	eventFilterConfig state_stream.EventFilterConfig
+	maxStreams        int32
+	streamCount       *atomic.Int32
```
```suggestion
	activeStreamCount       *atomic.Int32
```

durkmurder commented on engine/access/rest/routes/websocket_handler.go:
```diff
@@ -0,0 +1,279 @@
+	maxStreams        int32
+	streamCount       *atomic.Int32
+}
+
```
```suggestion

var _ http.Handler = (*WSHandler)(nil)

```

peterargue commented on engine/access/rest/routes/websocket_handler.go:
```diff
@@ -0,0 +1,283 @@
+		// 2) when an any additional message is received from the client
+		_, msg, err := wsController.conn.ReadMessage()
+		if err != nil {
+			return
```
after testing this out, I think it would be helpful to update `wsController.readChannel` to be receive errors, and then have each of these cases return the error. 

then in the `writeEvents` loop, return the error is one is received. This would make it easier for clients to debug.

peterargue reviewed (commented):
thanks for the updates! Added a few more, mostly nits and testing requests.

I tried to test this out using localnet and this curl command:
```
curl -vvv --include \
     --no-buffer \
     --header "Connection: Upgrade" \
     --header "Upgrade: websocket" \
     --header "Sec-WebSocket-Key: SGVsbG8sIHdvcmxkIQ==" \
     --header "Sec-WebSocket-Version: 13" \
     http://localhost:4004/v1/subscribe_events
```

it seems to timeout after a few seconds. I modified the code a bit to return errors from the read handler, and it returned
```
�>�read tcp 192.168.80.10:8070->192.168.80.1:37030: i/o timeout* Connection #0 to host localhost left intact
```

I suspect this is from the pong handler config.

UlyanaAndrukhiv commented on engine/access/rest/routes/subscribe_events_test.go:
```diff
@@ -0,0 +1,326 @@
+			respRecorder := NewHijackResponseRecorder()
+			// closing the connection after 5 seconds
+			go func() {
+				time.Sleep(5 * time.Second)
```
In current implementation, testHijackResponseRecorder waits for the connection to be closed before checking all actual responses. I understand that the 5-second delay is lengthy, so to address this, I've reduced the timeout to 1 second, which should significantly speed up the test process while still allowing us to verify all responses correctly.

UlyanaAndrukhiv commented:
> thanks for the updates! Added a few more, mostly nits and testing requests.
> 
> I tried to test this out using localnet and this curl command:
> 
> ```
> curl -vvv --include \
>      --no-buffer \
>      --header "Connection: Upgrade" \
>      --header "Upgrade: websocket" \
>      --header "Sec-WebSocket-Key: SGVsbG8sIHdvcmxkIQ==" \
>      --header "Sec-WebSocket-Version: 13" \
>      http://localhost:4004/v1/subscribe_events
> ```
> 
> it seems to timeout after a few seconds. I modified the code a bit to return errors from the read handler, and it returned
> 
> ```
> �>�read tcp 192.168.80.10:8070->192.168.80.1:37030: i/o timeout* Connection #0 to host localhost left intact
> ```
> 
> I suspect this is from the pong handler config.

Thank you for all your comments! 

The timeout issue you encountered related to the WebSocket ping-pong mechanism. Curl is not well-suited for WebSocket testing as it does not handle WebSocket ping-pong processes. I tried using curl with the Gorilla WebSocket '[chat](https://github.com/gorilla/websocket/tree/main/examples/chat)' example and received the same `Connection #0 to host localhost left intact` message.

I recommend `wscat` tool, which I have been using to test event streaming on localnet.

durkmurder commented on engine/access/rest/routes/router.go:
```diff
@@ -12,11 +12,20 @@ import (
 )
 
-func NewRouter(backend access.API, logger zerolog.Logger, chain flow.Chain, restCollector module.RestMetrics) (*mux.Router, error) {
+type RouterBuilder struct {
```
Can you add a small godoc to describe `RouterBuilder` and its functions?

durkmurder commented on engine/access/rest/routes/subscribe_events.go:
```diff
@@ -0,0 +1,33 @@
+func SubscribeEvents(
+	ctx context.Context,
+	request *request.Request,
+	wsController *WebsocketController) (state_stream.Subscription, error) {
```
```suggestion
	wsController *WebsocketController,
) (state_stream.Subscription, error) {
```

durkmurder commented on engine/access/rest/routes/subscribe_events_test.go:
```diff
@@ -0,0 +1,338 @@
+	}
+}
+
+func (s *SubscribeEventsSuite) TestSubscribeEvents() {
```
A short godoc for each test would be appreciated, just to get general idea of what we are dealing with

durkmurder commented on engine/access/rest/routes/subscribe_events_test.go:
```diff
@@ -0,0 +1,338 @@
+	require.Contains(t, recorder.responseBuff.String(), expected)
+}
+
+func requireResponse(t *testing.T, recorder *testHijackResponseRecorder, expected []*state_stream.EventsResponse) {
```
Same here, lets add a godoc which describes when response is valid

durkmurder commented on engine/access/rest/routes/test_helpers.go:
```diff
@@ -24,29 +30,111 @@ const (
+	closed       chan struct{}
+	responseBuff *bytes.Buffer
+}
+
```
```suggestion

var _ http.Hijacker = (*testHijackResponseRecorder)(nil)

```

durkmurder reviewed (approved):
Looks good from my side, good job.

peterargue commented on integration/tests/access/rest_state_stream_test.go:
```diff
@@ -0,0 +1,230 @@
+// events which received from grpc api
+func (s *RestStateStreamSuite) requireEvents(receivedEventsResponse []*state_stream.EventsResponse) {
+	// make sure there are received events
+	require.GreaterOrEqual(s.T(), len(receivedEventsResponse), 1, "expect received events")
```
this will pass as long as at least one response is returned, regardless of whether or not there were any errors. We should also require that there is at least one response with a non-empty set of events. 

Maybe you could keep a count in the inner `receivedEventMap` loop below, then assert at the end that the count is > 0?

peterargue reviewed (commented):
Looking good. I think this is ready after these last few comments.

peterargue reviewed (approved):
looks good. Great work @UlyanaAndrukhiv!
